pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,889 @@
+		webSocketConfig  config.WebSocketConfig
+		controllerConfig config.ControllerConfig
+		ctx              *context.Context
+		wantErr          bool
+	}{
+		{"TestEdgeHubController_initial: Proper input", EdgeHubController{
+			config: &config.ControllerConfig{
+				HeartbeatPeroid: 150 * time.Second,
+				RefreshInterval: 15 * time.Minute,
+				AuthInfosPath:   "/var/IEF/secret",
+				PlacementUrl:    TestServer.URL + "/proper_request",
+				ProjectID:       "foo",
+				NodeId:          "bar",
+			},
+			stopChan:   make(chan struct{}),
+			syncKeeper: make(map[string]chan model.Message),
+		}, config.WebSocketConfig{
+			CertFilePath: "/tmp/kubeedge/testData/edge.crt",
+			KeyFilePath:  "/tmp/kubeedge/testData/edge.key",
+		}, config.ControllerConfig{
+			PlacementUrl: TestServer.URL,
+			ProjectID:    "foo",
+			NodeId:       "bar",
+		},
+			context.GetContext(context.MsgCtxTypeChannel), false},
+
+		{"TestEdgeHubController_initial: Wrong placement URL", EdgeHubController{
+			config: &config.ControllerConfig{
+				HeartbeatPeroid: 150 * time.Second,
+				RefreshInterval: 15 * time.Minute,
+				AuthInfosPath:   "/var/IEF/secret",
+				PlacementUrl:    TestServer.URL + "/wrong_url",
+				ProjectID:       "foo",
+				NodeId:          "bar",
+			}, stopChan: make(chan struct{}),
+			syncKeeper: make(map[string]chan model.Message),
+		}, config.WebSocketConfig{
+			CertFilePath: "/tmp/kubeedge/testData/edge.crt",
+			KeyFilePath:  "/tmp/kubeedge/testData/edge.key",
+		}, config.ControllerConfig{
+			PlacementUrl: TestServer.URL,
+			ProjectID:    "foo",
+			NodeId:       "bar",
+		},
+			context.GetContext(context.MsgCtxTypeChannel), false},
+
+		{"TestEdgeHubController_initial: No project Id & node Id", EdgeHubController{
+			config: &config.ControllerConfig{
+				HeartbeatPeroid: 150 * time.Second,
+				RefreshInterval: 15 * time.Minute,
+				AuthInfosPath:   "/var/IEF/secret",
+				PlacementUrl:    TestServer.URL + "/proper_request",
+				ProjectID:       "foo",
+				NodeId:          "bar",
+			},
+			stopChan:   make(chan struct{}),
+			syncKeeper: make(map[string]chan model.Message),
+		}, config.WebSocketConfig{
+			CertFilePath: "/tmp/kubeedge/testData/edge.crt",
+			KeyFilePath:  "/tmp/kubeedge/testData/edge.key",
+		}, config.ControllerConfig{
+			PlacementUrl: TestServer.URL,
+			ProjectID:    "",
+			NodeId:       "",
+		},
+			context.GetContext(context.MsgCtxTypeChannel), false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			edgeHubConfig := config.GetConfig()
+			edgeHubConfig.WSConfig = tt.webSocketConfig
+			edgeHubConfig.CtrConfig = tt.controllerConfig
+			ehc := &EdgeHubController{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if err := ehc.initial(tt.ctx); (err != nil) != tt.wantErr {
```

Comment:
instead of just checking if err is nil or not, can we not assert the error ? This will make sure that the failure scenario for which the test case is designed for is achieved. 
Applicable to similar cases below